aten/src/ATen/native/cuda/Sort.cu

```diff
@@ -85,6 +265,33 @@ std::tuple<Tensor &,Tensor &> sort_out_stable_cuda(const Tensor & self, c10::opt
 
   TORCH_CHECK(nsort <= std::numeric_limits<int>::max(),
     "The dimension being sorted can not have more than INT_MAX elsments.");
+  // use inplace algorithm for smaller input sizes without stable=True
+  if (should_use_small_sort(self, dim) && !stable.value()) {
+    // from thc: sorted->values, indices->indices, input->self
+    TensorArg self_arg{self, "self", 1}, values_arg{values, "values", 2}, indices_arg{indices, "indices", 3};
+    checkAllSameGPU("small_sort", {self_arg, values_arg, indices_arg});
+    at::maybe_wrap_dim(dim, self);
+
+    if (!values.defined()) {
+      values = at::empty_like(self, self.options());
+    }
+    if (!indices.defined()) {
+      indices = at::empty_like(self, self.options().dtype(kLong));
+    }
+
+    // Make sure sufficient output space is allocated
+    std::vector<int64_t> self_size = self.sizes().vec();
+    at::native::resize_output(values, self_size);
+    at::native::resize_output(indices, self_size);
+    fillSliceWithIndex(indices, dim);
+
+    // We sort k/v pairs in-place; copy unsorted input to output
+    values.copy_(self);
+
+    // Sort using our in-place k/v kernel that supports arbitrary
+    // layout
+    sortKeyValueInplace(values, indices, dim, descending);
```

Comment:
you should return here, before going through generic implementation?

Reply:
oof, good catch!